File: packages/dcf_primitives/ios/Classes/Utilities/DCFViewportObserver.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit
import dcflight

/// Low-level viewport detection system - similar to web IntersectionObserver
/// Any view can register for viewport visibility callbacks
class DCFViewportObserver {
    static let shared = DCFViewportObserver()
    
    private var observedViews: [UIView: ViewportConfig] = [:]
    private var scrollViewObservers: [UIScrollView: Set<UIView>] = [:]
    
    private init() {}
    
    /// Register a view for viewport detection
    func observe(_ view: UIView, config: ViewportConfig) {
        observedViews[view] = config
        
        // Find parent scroll view
        if let scrollView = findParentScrollView(view) {
            if scrollViewObservers[scrollView] == nil {
                scrollViewObservers[scrollView] = Set<UIView>()
                setupScrollObserver(scrollView)
            }
            scrollViewObservers[scrollView]?.insert(view)
        } else {
            // No scroll view - check initial visibility
            checkVisibility(view, config: config)
        }
    }
    
    /// Unregister a view
    func unobserve(_ view: UIView) {
        observedViews.removeValue(forKey: view)
        
        // Remove from scroll view observers
        for (scrollView, views) in scrollViewObservers {
            if views.contains(view) {
                scrollViewObservers[scrollView]?.remove(view)
                if scrollViewObservers[scrollView]?.isEmpty == true {
                    scrollViewObservers.removeValue(forKey: scrollView)
                }
            }
        }
    }
    
    /// Find parent scroll view
    private func findParentScrollView(_ view: UIView) -> UIScrollView? {
        var current: UIView? = view.superview
        while current != nil {
            if let scrollView = current as? UIScrollView {
                return scrollView
            }
            current = current?.superview
        }
        return nil
    }
    
    /// Setup scroll observer for a scroll view
    private func setupScrollObserver(_ scrollView: UIScrollView) {
        // Listen for scroll notifications
        NotificationCenter.default.addObserver(
            forName: NSNotification.Name("DCFScrollViewDidScroll"),
            object: scrollView,
            queue: .main
        ) { [weak self] _ in
            self?.checkViewsInScrollView(scrollView)
        }
        
        // Also check on initial setup
        DispatchQueue.main.async { [weak self] in
            self?.checkViewsInScrollView(scrollView)
        }
    }
    
    /// Check visibility of views in a scroll view
    private func checkViewsInScrollView(_ scrollView: UIScrollView) {
        guard let views = scrollViewObservers[scrollView] else { return }
        
        for view in views {
            if let config = observedViews[view] {
                checkVisibility(view, inScrollView: scrollView, config: config)
            }
        }
    }
    
    /// Check if view is visible (no scroll view)
    private func checkVisibility(_ view: UIView, config: ViewportConfig) {
        guard let window = view.window else { return }
        
        let viewFrame = view.convert(view.bounds, to: window)
        let windowBounds = window.bounds
        
        let isVisible = viewFrame.intersects(windowBounds)
        let intersectionRatio = calculateIntersectionRatio(viewFrame, windowBounds)
        
        handleVisibilityChange(view, isVisible: isVisible, ratio: intersectionRatio, config: config)
    }
    
    /// Check if view is visible in scroll view
    private func checkVisibility(_ view: UIView, inScrollView scrollView: UIScrollView, config: ViewportConfig) {
        let scrollViewBounds = scrollView.bounds
        let viewFrame = view.convert(view.bounds, to: scrollView)
        
        let isVisible = viewFrame.intersects(scrollViewBounds)
        let intersectionRatio = calculateIntersectionRatio(viewFrame, scrollViewBounds)
        
        handleVisibilityChange(view, isVisible: isVisible, ratio: intersectionRatio, config: config)
    }
    
    /// Calculate intersection ratio (0.0 to 1.0)
    private func calculateIntersectionRatio(_ viewFrame: CGRect, _ containerBounds: CGRect) -> CGFloat {
        let intersection = viewFrame.intersection(containerBounds)
        if intersection.isNull {
            return 0.0
        }
        
        let viewArea = viewFrame.width * viewFrame.height
        if viewArea == 0 {
            return 0.0
        }
        
        let intersectionArea = intersection.width * intersection.height
        return intersectionArea / viewArea
    }
    
    /// Handle visibility change
    private func handleVisibilityChange(_ view: UIView, isVisible: Bool, ratio: CGFloat, config: ViewportConfig) {
        // Check threshold
        let threshold = config.amount ?? 0.0
        let meetsThreshold = ratio >= CGFloat(threshold)
        
        // Get current state
        let wasVisible = objc_getAssociatedObject(view, UnsafeRawPointer(bitPattern: "viewportVisible".hashValue)!) as? Bool ?? false
        
        if isVisible && meetsThreshold && !wasVisible {
            // Entered viewport
            objc_setAssociatedObject(view, UnsafeRawPointer(bitPattern: "viewportVisible".hashValue)!, true, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            propagateEvent(on: view, eventName: "onViewportEnter", data: [
                "intersectionRatio": ratio,
                "isIntersecting": true
            ])
        } else if (!isVisible || !meetsThreshold) && wasVisible {
            // Left viewport
            if !config.once {
                objc_setAssociatedObject(view, UnsafeRawPointer(bitPattern: "viewportVisible".hashValue)!, false, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
                propagateEvent(on: view, eventName: "onViewportLeave", data: [
                    "intersectionRatio": ratio,
                    "isIntersecting": false
                ])
            }
        }
    }
}

struct ViewportConfig {
    let once: Bool
    let amount: Double? // 0.0 to 1.0
}

File: packages/dcf_primitives/android/src/main/kotlin/com/dotcorr/dcf_primitives/utils/DCFViewportObserver.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcf_primitives.utils

import android.graphics.Rect
import android.view.View
import android.view.ViewGroup
import androidx.core.widget.NestedScrollView
import com.dotcorr.dcflight.components.DCFTags
import com.dotcorr.dcflight.components.propagateEvent
import com.dotcorr.dcf_primitives.components.DCFScrollableView

/**
 * Low-level viewport detection system - similar to web IntersectionObserver
 * Any view can register for viewport visibility callbacks
 */
object DCFViewportObserver {
    private val observedViews = mutableMapOf<View, ViewportConfig>()
    private val scrollViewObservers = mutableMapOf<ViewGroup, MutableSet<View>>()
    
    /**
     * Register a view for viewport detection
     */
    fun observe(view: View, config: ViewportConfig) {
        observedViews[view] = config
        
        // Find parent scroll view
        val scrollView = findParentScrollView(view)
        if (scrollView != null) {
            if (!scrollViewObservers.containsKey(scrollView)) {
                scrollViewObservers[scrollView] = mutableSetOf()
                setupScrollObserver(scrollView)
            }
            scrollViewObservers[scrollView]?.add(view)
        } else {
            // No scroll view - check initial visibility
            checkVisibility(view, config)
        }
    }
    
    /**
     * Unregister a view
     */
    fun unobserve(view: View) {
        observedViews.remove(view)
        
        // Remove from scroll view observers
        scrollViewObservers.forEach { (scrollView, views) ->
            if (views.contains(view)) {
                views.remove(view)
                if (views.isEmpty()) {
                    scrollViewObservers.remove(scrollView)
                }
            }
        }
    }
    
    /**
     * Find parent scroll view
     */
    private fun findParentScrollView(view: View): ViewGroup? {
        var current: View? = view.parent as? View
        while (current != null) {
            if (current is NestedScrollView || current is DCFScrollableView) {
                return current as ViewGroup
            }
            current = current.parent as? View
        }
        return null
    }
    
    /**
     * Setup scroll observer for a scroll view
     */
    private fun setupScrollObserver(scrollView: ViewGroup) {
        if (scrollView is NestedScrollView) {
            scrollView.viewTreeObserver.addOnScrollChangedListener {
                checkViewsInScrollView(scrollView)
            }
        }
    }
    
    /**
     * Check visibility of views in a scroll view
     */
    private fun checkViewsInScrollView(scrollView: ViewGroup) {
        val views = scrollViewObservers[scrollView] ?: return
        
        views.forEach { view ->
            observedViews[view]?.let { config ->
                checkVisibility(view, inScrollView = scrollView, config)
            }
        }
    }
    
    /**
     * Check if view is visible (no scroll view)
     */
    private fun checkVisibility(view: View, config: ViewportConfig) {
        val windowRect = Rect()
        view.getGlobalVisibleRect(windowRect)
        
        val isVisible = windowRect.width() > 0 && windowRect.height() > 0
        val intersectionRatio = if (isVisible) {
            val viewArea = view.width * view.height
            if (viewArea > 0) {
                (windowRect.width() * windowRect.height()).toDouble() / viewArea
            } else {
                0.0
            }
        } else {
            0.0
        }
        
        handleVisibilityChange(view, isVisible, intersectionRatio, config)
    }
    
    /**
     * Check if view is visible in scroll view
     */
    private fun checkVisibility(view: View, inScrollView: ViewGroup, config: ViewportConfig) {
        val scrollViewRect = Rect()
        inScrollView.getGlobalVisibleRect(scrollViewRect)
        
        val viewRect = Rect()
        view.getGlobalVisibleRect(viewRect)
        
        val intersection = Rect()
        val isVisible = intersection.setIntersect(scrollViewRect, viewRect)
        val intersectionRatio = if (isVisible) {
            val viewArea = view.width * view.height
            if (viewArea > 0) {
                (intersection.width() * intersection.height()).toDouble() / viewArea
            } else {
                0.0
            }
        } else {
            0.0
        }
        
        handleVisibilityChange(view, isVisible, intersectionRatio, config)
    }
    
    /**
     * Calculate intersection ratio (0.0 to 1.0)
     */
    private fun calculateIntersectionRatio(viewRect: Rect, containerRect: Rect): Double {
        val intersection = Rect()
        val intersects = intersection.setIntersect(viewRect, containerRect)
        
        if (!intersects) {
            return 0.0
        }
        
        val viewArea = viewRect.width() * viewRect.height()
        if (viewArea == 0) {
            return 0.0
        }
        
        val intersectionArea = intersection.width() * intersection.height()
        return intersectionArea.toDouble() / viewArea
    }
    
    /**
     * Handle visibility change
     */
    private fun handleVisibilityChange(view: View, isVisible: Boolean, ratio: Double, config: ViewportConfig) {
        // Check threshold
        val threshold = config.amount ?: 0.0
        val meetsThreshold = ratio >= threshold
        
        // Get current state
        val wasVisible = view.getTag(DCFTags.VIEWPORT_VISIBLE_KEY) as? Boolean ?: false
        
        if (isVisible && meetsThreshold && !wasVisible) {
            // Entered viewport
            view.setTag(DCFTags.VIEWPORT_VISIBLE_KEY, true)
            propagateEvent(view, "onViewportEnter", mapOf(
                "intersectionRatio" to ratio,
                "isIntersecting" to true
            ))
        } else if ((!isVisible || !meetsThreshold) && wasVisible) {
            // Left viewport
            if (!config.once) {
                view.setTag(DCFTags.VIEWPORT_VISIBLE_KEY, false)
                propagateEvent(view, "onViewportLeave", mapOf(
                    "intersectionRatio" to ratio,
                    "isIntersecting" to false
                ))
            }
        }
    }
}

data class ViewportConfig(
    val once: Boolean = false,
    val amount: Double? = null // 0.0 to 1.0
)

